woocommerce/storefront
# reinstate phpcs php linting

Prompted by #1393
Fixes #1356

Update: woah cool there is an issue for this! #1356 🎩 @nielslange 

The original goal of this PR was to update the WordPress coding standards PHPcs ruleset, and anything necessary related to it.

I discovered that our phpcs setup wasn't in working order, so have upgraded and replaced various packages. 

~~Work in progress - need to ensure we have enough rules in place that match the existing code. Currently have lots of violations.~~ All violations now fixed and pre-commit should catch issues in new code.

Summary of what's included in this PR:

- Updated and set versions for various phpcs related packages. Key outcome here is getting on the latest WordPress standards. fac37e55bfb523c07e37299edf46946839290d5f
- Added/tweaked commands in composer.json and package.json for visibility and to encourage use/familiarity.
- Set min PHP version of sniffer config to match plugin min PHP version.
- Mega commit fixing all formatting issues automatically using `phpcbf`. 7c90d2311b0433ee9cb0f9d896704c64db3ae0e5
- Manual fixing of a few more interesting violations (~10 issues). One commit per file for visibility. Note this included:
  - Adding/fixing version param in all uses of `wp_enqueue_style`. This means that various third-party assets will be cached less aggressively (now tied to `$storefront_version`).
  - Missing escaping of comment dates in `storefront_comment()` (🧤)
  - Use `gmdate()` instead of ambiguous `date()`.
- Fixed one rule ignore comment to latest format. (Still ignoring, didn't investigate why)

##### We still have some warnings throughout code
> Also notice that when I run that command I get several warnings. I assume that's expected? – @Aljullu 

I've fixed all errors but have not touched any warnings. Warnings should be _carefully_ handled as follow up - linter warnings can be very subtle/tricky/risky  🐲 to fix, so I'm not rushing to fix them all in this PR!

**Update** I have bulk-fixed an older style of ignore comment to reduce the amount of open warnings. These comments opt-out of XSS checks, and we have a lot of them. c2b82fac62038925df1db0bf0fb6f7ed555808c9

- Old comment `// WPCS: XSS ok.`
- Current comment, using `phpcs:ignore` `// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped`


##### Travis / CI is _not_ included (yet)
Note there's nothing in this PR to trigger linter check as a CI step (e.g. on PRs). We should do that next! #1395 

### How to test the changes
#### Test linter tooling
1. Get on this branch, `composer install`, `npm install`.
   - Note: `composer install` is what sets up the `pre-commit` hook (TIL).
2. Run linter `npm run lint:php`. Should pass with flying colours (* and plenty of warnings 😿).
3. Write some [bad PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/).
4. Run linter again - should tell you what's wrong.
5. Try to commit the bad changes - commit hook should catch it. 
5. Write some PHP with auto-fixable changes (e.g. spaces/tabs, dangling commas, long lines, multi-line func calls).
6. Run auto fix to `npm run lint:php:fix`. Should fix the errors so you can commit.

#### Broad smoke testing of a cross-section of Storefront functionality
As this is a big change, any testing of Storefront features would be appreciated. 

cc @woocommerce/rubik I've tagged you all for review – first in first served. I don't expect everyone to review – grateful for any reviews or testing.

If you could run this PR (or master after merge) in your dev environment and keep an eye out for issues I would really appreciate it 🙌 🙇 

#### Affected flows
- Everything (in theory). 
- Comment dates now escaped. 

### Changelog

> Dev – Reinstate phpcs PHP linting. #1463
